### PR TITLE
add plugin subsonicupdate auth parameter

### PR DESCRIPTION
**Pull request by vincentDcmps**
## Description

this modification will permit that subsonicupdate works with subsonic fork like https://github.com/spl0k/supysonic

### CODE REVIEW

File: beetsplug/subsonicupdate.py
Comment (wisp3rwind) [inline beetsplug/subsonicupdate.py:98]: This should be `config['subsonic']['auth']`, right?

File: beetsplug/subsonicupdate.py
Comment (wisp3rwind) [inline beetsplug/subsonicupdate.py:110]: Is this whitespace change intentional? Personally, I'd say this block of code is sufficiently long to benefit from the additional linebreak.

File: docs/plugins/subsonicupdate.rst
Comment (wisp3rwind) [inline docs/plugins/subsonicupdate.rst:33]: I think this needs to be expanded: For me, this sounds as if the value should be the token or password, rather than just an indication of the authentication method. The valid choices (`token` and `enc`) should be indicated, and the description of the `pass` key below should reflect which values it can take in each case.

Since the `enc:` prefix seems to be allowed for tokens, too, maybe it would be better to rename this option to `plain`?

File: beetsplug/subsonicupdate.py
Comment (wisp3rwind) [inline beetsplug/subsonicupdate.py:93]: Is this valid for all versions of subsonic?

**Review by wisp3rwind — CHANGES_REQUESTED** [review]
I left a few comments on the code. I'm not familiar with subsonic itself, though, so I can't really comment on whether these changes are sensible. @jef you have worked on this plugin most recently, could you have a look?

File: beetsplug/subsonicupdate.py
Comment (vincentDcmps) [reply beetsplug/subsonicupdate.py]: correction done thanks

File: beetsplug/subsonicupdate.py
Comment (vincentDcmps) [reply beetsplug/subsonicupdate.py]: linebreak rollback

File: beetsplug/subsonicupdate.py
Comment (vincentDcmps) [reply beetsplug/subsonicupdate.py:93]: I have test with  http://demo.subsonic.org/ seem works

File: docs/plugins/subsonicupdate.rst
Comment (vincentDcmps) [reply docs/plugins/subsonicupdate.rst]: change enc by plain

File: docs/plugins/subsonicupdate.rst
Comment (jef) [inline docs/plugins/subsonicupdate.rst:33]:
```suggestion
- **auth**: Authentication method. Default: ``token``. Use ``enc`` for APIs <= 1.12.0.
```

**Review by jef — CHANGES_REQUESTED** [review]
I have a few comments here.

First of all, thanks for the contribution! This is great! Secondly, I think we could do something interesting here that could build other deprecated API rather than using `auth = token`.

I think if we call `/rest/ping` and extract the `version` (see http://demo.subsonic.org/rest/ping), we can call the proper API. That way the user doesn't get confused on what to use with `auth`.

This is not necessary by any means, but I think it could be interesting and wanted to bring it up.

I think we should go two routes:

a. review the comments and update some of the verbiage
b. implement the API version check

Feel free to choose any routes and if you have any concerns, let me know ☺️

Again, thank you very much for this!

File: beetsplug/subsonicupdate.py
Comment (vincentDcmps) [reply beetsplug/subsonicupdate.py:23]: OK so I think in waiting to do the ping method best way is to put simply token or password

File: beetsplug/subsonicupdate.py
Comment (vincentDcmps) [reply beetsplug/subsonicupdate.py:48]: to not break current config default must be to token, currently plugins use only token authent

File: beetsplug/subsonicupdate.py
Comment (jef) [reply beetsplug/subsonicupdate.py:23]: Awesome! I look forward to seeing it. Thanks so much for your time :)

Let me know if you need anything in the meantime.

**Comment by vincentDcmps** [comment]
@jef I have implemented the version check like sugest

File: beetsplug/subsonicupdate.py
Comment (jef) [inline beetsplug/subsonicupdate.py:102]:
```suggestion
    def __get_version(self):
```

I believe we want to make this private.

File: beetsplug/subsonicupdate.py
Comment (jef) [inline beetsplug/subsonicupdate.py:103]:
```suggestion
        url = self.__format_url("ping")
```

Docs have `.view`, but I don't believe we need `.view`. This is also for the `starScan.view`.

**Review by jef — COMMENTED** [review]
Awesome! Things look great to me. Just one last thing and I think we're good to go!

**Comment by jef** [comment]
@vincentDcmps thanks for all the hard work on this, this looks great! This will also help us in the future with other API calls if we plan to have any!

File: beetsplug/subsonicupdate.py
Comment (vincentDcmps) [reply beetsplug/subsonicupdate.py:103]: like I say before I have put that way because .view work with classic subsonic but not with supysonic
 I will contact maintainer to know if we can add endpoint without .view. so for me it's better to let .view for now no?

File: beetsplug/subsonicupdate.py
Comment (jef) [reply beetsplug/subsonicupdate.py]: Ahh, it's ok! Luckily it works in both. I didn't realize this was also another deprecated API thing.

Thank you!

File: beetsplug/subsonicupdate.py
Comment (jef) [reply beetsplug/subsonicupdate.py:103]: Sorry, I didn't see your other comment.

No that's fine. Since `.view` is compatible with both versions, I think this is fine.

Thank you very much ☺️

**Review by jef — APPROVED** [review]
Looks good to me! Thank for for all your hard work and working with me on this!

**Comment by jef** [comment]
@wisp3rwind, ready to go. Can you give it a once over?

File: test/test_subsonicupdate.py
Comment (jef) [inline test/test_subsonicupdate.py:52]:
```suggestion
        "status": "failed",
```

**Review by jef — APPROVED** [review]

File: beetsplug/subsonicupdate.py
Comment (sampsyo) [inline beetsplug/subsonicupdate.py:37]: FWIW, it's generally nice for module-level constants to get `CAPITAL_NAMES`, like `AUTH_TOKEN_VERSION`.

**Review by sampsyo — COMMENTED** [review]
Looking great! One tiny style suggestion within.

It also seems a little odd to compare version strings lexically, as opposed to doing `tuple(version.split('.')) > (1, 12)` to do numerical comparisons, but I can't actually concoct a reasonable example where this does the wrong thing, so maybe it's fine.

File: beetsplug/subsonicupdate.py
Comment (wisp3rwind) [inline beetsplug/subsonicupdate.py:80]: Maybe this wording is clearer?
```suggestion
        """Get the Subsonic URL to trigger the given endpoint.
```

**Review by wisp3rwind — APPROVED** [review]
Looking great! Thanks @vincentDcmps and @jef for the great work, the version check is a lot nicer than yet another config option!

I added some more nitpicks in the inline comments in addition to what @sampsyo said. None of them are really blocking the merge here. Please indicate if you'd rather be done with this PR, then I'll push those changes myself.

File: beetsplug/subsonicupdate.py
Comment (wisp3rwind) [inline beetsplug/subsonicupdate.py:57]: Minor rewording and style suggestion:
```suggestion
        if self.version > authtokenversion:
            self.auth = "token"
        else:
            self.auth = "password"
        self._log.info(u"using '{}' authentication method"
                .format(self.auth))
```

Is `> 1.12` the right version check, or should it rather be `>= 1.13`?

Comment (jef) [reply beetsplug/subsonicupdate.py:57]: Should be `<= 1.12`, so `> 1.12` is fine!

Comment (wisp3rwind) [reply beetsplug/subsonicupdate.py:57]: I'm asking because the API docs state that token authentication is new in `1.13.0`. There doesn't seem to be a version `1.12.1`, so this should indeed work. I don't know subsonic's versioning and release scheme, so no idea whether there's a chance that a `1.12.1` API version will ever be released (probably not, since `1.16.1` is current).

**Comment by vincentDcmps** [comment]
Hi everyone I just push your last suggestion. 
just the string comparison let as before because I find ```tuple(int(s) for s in version.split('.')) > (1, 12)``` less Human readable but it's a point of view ;)

**Comment by wisp3rwind** [comment]
> Hi everyone I just push your last suggestion.
> just the string comparison let as before because I find `tuple(int(s) for s in version.split('.')) > (1, 12)` less Human readable but it's a point of view ;)

Actually, looking at this again, lexical comparison of strings is incorrect:
```
>>> "1.13" > "1.12"
True
>>> "1.12" > "1.12"
False
>>> "1.11" > "1.12"
False
>>> "1.6" > "1.12"
True
```
note the last example.

**Comment by sampsyo** [comment]
Good find! Not sure why I couldn't think of that when I was trying to cook up examples before. :smiley:

**Comment by vincentDcmps** [comment]
OK I admit defeat ;) change done

**Comment by wisp3rwind** [comment]
> OK I admit defeat ;) change done

Thanks! The code is really shaping up now :tada:

I did spot another issue with the version check, though:
- If an error occurs in `__get_version`, it returns nothing (i.e. it implicitly returns `None`). That's fine I would say, since `None` transports the notion of "can't determine the version" sufficiently well. It might be good for readability to make the `return None` explicit. However, the next thing that happens is the comparison
```
>>> None > (1, 12)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '>' not supported between instances of 'NoneType' and 'tuple'
```
that is, beets crashes with an uncaught exception. It would be much better to exit without a traceback (`__get_version` already prints a diagnostic message, after all), or maybe even ignore the error, since apart from the subsonic update, an import session could still succeed. This is not an edge case that we could ignore, since this can fail for reasons such as the server being down.

- The version check is done too eagerly, namely during plugin initialization, which always happens on beets startup. However, the plugin does nothing unless `beet import` is being run. So the version check might be unnecessary, or worse could crash beets even if running some other command. The latter case might be remediated by fixing the first bullet point above. Still, we should not be contacting the server needlessly. So the version should only be queried when authentication needs to be done for the first time, which is in `start_scan`.
It might be easiest to just move the corresponding code from `__init__` to that function. Another option would be to turn `version` and `auth` into [properties](https://docs.python.org/3.9/library/functions.html#property) for which the getter functions contact the server when they're called for the first time. Right now, I don't see much of a difference between both methods. The solution using properties might be useful should the plugin ever gain additional functionality.

Let us know if you need some guidance for these changes!

**Review by wisp3rwind — CHANGES_REQUESTED** [review]
Loading the plugin is safe now :+1:

I do think that the point that I raised about the version check being too eager should be addressed, too. One way to do so would be to move the version check to `start_scan`, and in that function simply exit early if the version check fails. That should result in the same overall behaviour as your current code (i.e. just log an error if unable to determine the version, but not abort the import session).

The issue is that if you run e.g. `beet ls <some query>` the plugin will still be loaded and attempt to connect to the subsonic server to determine the version. I really think that `beet ls` should not cause any network traffic, though. In addition, I guess this can delay such commands quite a bit if the server is slow to respond.

**Review by jef — APPROVED** [review]
Great use of `@property` in my opinion! I think this looks pretty good to me. What do you think @wisp3rwind?

Thank for all the hard work here @vincentDcmps, I appreciate you sticking around and seeing this through.

**Review by wisp3rwind — APPROVED** [review]

**Comment by wisp3rwind** [comment]
Looks good to me :tada: If you want, you could add your credits to the [changelog](https://github.com/beetbox/beets/blob/master/docs/changelog.rst). Otherwise, I'm going to merge this soon.

**Comment by vincentDcmps** [comment]
changelog updated, thanks for your support.
if you need beginners help you are welcome ;)

**Comment by wisp3rwind** [comment]
> changelog updated, thanks for your support.
> if you need beginners help you are welcome ;)

By all means, please stay around and go on improving beets :)

**Comment by jef** [comment]
> if you need beginners help you are welcome ;)

Better than me, I barely use Python 😆 Thanks for all the help!

**Pull request merged**
